### PR TITLE
Pull request for basic distance and overlap methods between sources

### PR DESCRIPTION
**Pull request by logang**
added pull request for changes to extraction/source.py that include basic distance metrics for sources

### CODE REVIEW

File: python/thunder/extraction/source.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 from numpy import asarray, mean, sqrt, ndarray, amin, amax, concatenate, sum, zeros, maximum, \
     argmin, newaxis, ones, delete, NaN, inf, isnan
 
+import numpy.linalg as la
+from scipy.stats import spearmanr
```
Comment (freeman-lab) [inline python/thunder/extraction/source.py]: This looks like it's unused

File: python/thunder/extraction/source.py [hunk 2]
```diff
@@ -106,18 +109,66 @@ def restore(self, skip=None):
...
+            self.coordinates = self.coordinates.tolist()
+        if isinstance(other.coordinates, ndarray):
+            other.coordinates = other.coordinates.tolist()
+        intersection = [a for a in self.coordinates if a in other.coordinates]
+        complement = [a for a in self.coordinates if a not in intersection]
+        if method == 'support':
+            hits = len(intersection)
+            misses = len(complement)
+            if counts:
+                return (hits,misses)
```
Comment (freeman-lab) [inline python/thunder/extraction/source.py]: insert space before `misses`

File: python/thunder/extraction/source.py [hunk 3]
```diff
@@ -1,6 +1,9 @@
 from numpy import asarray, mean, sqrt, ndarray, amin, amax, concatenate, sum, zeros, maximum, \
     argmin, newaxis, ones, delete, NaN, inf, isnan
 
+import numpy.linalg as la
```
Comment (freeman-lab) [inline python/thunder/extraction/source.py]: Can we move this one so it's locally scoped to `distance`, and use `from numpy.linalg import norm`?

File: python/thunder/extraction/source.py [hunk 4]
```diff
@@ -106,18 +109,66 @@ def restore(self, skip=None):
...
+        """
+        Compute the overlap between current source and 'other' source, in terms 
+        of either support or similarity of coefficients. 
+
+        Parameters
+        ----------
+        method : str
+            Compare either support of source coefficients ('support'), or the 
+            source spatial filters (not yet implemented).
+
```
Comment (freeman-lab) [inline python/thunder/extraction/source.py]: Remove empty line

File: python/thunder/extraction/source.py [hunk 5]
```diff
@@ -106,18 +109,66 @@ def restore(self, skip=None):
...
+            #TODO: this is actually a bit complicated, as support size can matter... we
+            # could use a fixed box size or the union of source supports?
+            raise NotImplementedError()
+        else:
+            raise ValueError('Given method is not supported.')
 
     def tolist(self):
         """
         Convert array-like attributes to list
+
```
Comment (freeman-lab) [inline python/thunder/extraction/source.py]: Remove empty line

File: python/thunder/extraction/source.py [hunk 6]
```diff
@@ -106,18 +109,66 @@ def restore(self, skip=None):
...
+        intersection = [a for a in self.coordinates if a in other.coordinates]
+        complement = [a for a in self.coordinates if a not in intersection]
+        if method == 'support':
+            hits = len(intersection)
+            misses = len(complement)
+            if counts:
+                return (hits,misses)
+            else:
+                return hits/float(hits+misses)
+        elif method == 'corr':
```
Comment (freeman-lab) [inline python/thunder/extraction/source.py]: Or intersection? Specifically, take the intersection computed above, and compute a correlation coefficient on just those values?

**Comment by freeman-lab** [comment]
Thanks @logang ! This is looking really good. I left some minor comments, a couple additional things:

- Can you figure out why the tests are failing? To run them locally, just follow these [instructions](http://thunder-project.org/thunder/docs/contributing.html#running-the-tests). Something's broken in the distance calculation, but I can't quite tell how. It might be changing the default from `euclidean` to `l1`, in which case, we'll need to change the tests, or change the default. I think I have a slight preference for restoring `euclidean` as the default.

- I'd like to add something for the `corr` method before merging, rather than leave as a todo.

- In your PR we've lost the `similarity` function, which was designed to estimate the fraction of sources in one model also found in another, where "found in" is defined as having a value less than some threshold for one of the comparison operations. Can we add that back in? Maybe you found the name confusing? Open to alternatives. It should probably look like the one from before, but now include both `distance` and `overlap` as possible arguments for `metric`, which will be great to have!

**Comment by freeman-lab** [comment]
@logang just pinging you on this one =) would be awesome to merge in soon, as we've made a bunch of progress over on [neurofinder](https://github.com/CodeNeuro/neurofinder) and would love to use these metrics in the initial testing stages!

File: python/thunder/extraction/source.py (hunk 6 above)
Comment (logang) [inline python/thunder/extraction/source.py]: ok so we will do: (\rho*hits)/(hits+misses) where \rho is the spearman correlation

**Comment by freeman-lab** [comment]
Awesome, LGTM, thanks for the contribution @logang !

**Pull request merged**